### PR TITLE
Add shortcut to find out if an AbstractArray is differentiable

### PR DESCRIPTION
**Pull request by HGSilveri**
- Adds `AbstractArray.is_differentiable`, which checks if the stored array is a tensor with `requires_grad=True`

### CODE REVIEW

File: pulser-core/pulser/math/abstract_array.py [hunk 1]
```diff
@@ -71,6 +71,11 @@ def is_tensor(self) -> bool:
         """Whether the stored array is a tensor."""
         return self.has_torch() and isinstance(self._array, torch.Tensor)
 
+    @property
+    def is_differentiable(self) -> bool:
```
Comment (a-corni) [inline pulser-core/pulser/math/abstract_array.py:75]: Wouldn't `requires_grad` or `has_grad` be a more generic name for this property ? I have the feeling it does not only apply to differentiability (differentiability requires_grad, but requires_grad can be used for other perspectives, right ?)

File: pulser-core/pulser/math/abstract_array.py [hunk 2]
```diff
@@ -271,10 +276,7 @@ def __setitem__(self, indices: Any, values: AbstractArrayLike) -> None:
                 self._process_indices(indices)
             ] = values  # type: ignore[assignment]
         except RuntimeError as e:
-            if (
-                self.is_tensor
-                and cast(torch.Tensor, self._array).requires_grad
-            ):
+            if self.is_differentiable:
```
Comment (a-corni) [inline pulser-core/pulser/math/abstract_array.py:279]: Here for instance I have the feeling that self.requires_grad would be more suited...

**Review by a-corni — COMMENTED** [review]
It really looks good to me, but I would really much prefer having this property be named `requires_grad`, because that's how we use it in all our tests. However, if having `is_differentiable` is needed to improve the user experience, I suggest we have two properties `requires_grad` and `is_differentiable`, with `is_differentiable` simply returning `requires_grad`.

File: pulser-core/pulser/math/abstract_array.py (hunk 1 above)
Comment (HGSilveri) [reply pulser-core/pulser/math/abstract_array.py:75]: I used `is_differentiable` because I felt it was more explicit. `requires_grad` is clear for people familiar with torch but I find it quite opaque otherwise. 
Also, to me being differentiable = having a gradient. What do you have in mind that would fall outside of this scope?

Comment (a-corni) [reply pulser-core/pulser/math/abstract_array.py:75]: `is_differentiable` is definitely explicit. My only point is what this function does is what a function `requires_grad`  should do. We could imagine that at some point, `is_differentiable` can be different that `requires_grad` (might be conditions on the gradient, idk), I would prefer having two properties, one `requires_grad` and one `is_differentiable`, with `is_differentiable` currently returning the `requires_grad` property (I think that some people will still try to check the differentiability using `requires_grad`).

But it's very minor ;)

Comment (HGSilveri) [reply pulser-core/pulser/math/abstract_array.py:75]: Tbh I think it's overkill to bloat the source code with two ways to access the same property, at least at this stage. If you feel like `requires_grad` is better, I'll replace it

**Review by a-corni — APPROVED** [review]
LGTM ! Thanks :)
